File: kmp/nip41/src/commonMain/kotlin/com/appollo41/nip41/Derivation.kt
package com.appollo41.nip41

import fr.acinq.secp256k1.Secp256k1

/**
 * NIP-41 "Identity chain": derive the internal private key for generation `i`
 * from the root secret.
 *
 * `p_internal[i] = HKDF-SHA256(IKM=root_secret, salt="nip41-key-rotation-v1",
 *                              info="internal-key" || u32be(i), L=32)`
 *
 * If the reduced scalar is 0 or ≥ n (negligibly rare, ~2^-128) the spec says
 * to retry with one extra counter byte appended to `info`: `info || u8(counter)`,
 * starting at counter=1 and incrementing.
 *
 * [rootSecret] is read by HKDF and not retained or mutated; the caller may
 * zero its buffer after this call returns. The returned 32-byte scalar is a
 * fresh buffer the caller owns and is responsible for zeroing when done.
 */
public fun nip41InternalSeckey(rootSecret: ByteArray, generation: Int): ByteArray =
    deriveScalarAtCounter(rootSecret, generation, counterStart = 0).first

/**
 * Per-counter derivation primitive. Searches counter values `[counterStart, 255]`
 * and returns the first one whose HKDF output is a valid secp256k1 scalar (i.e.
 * `1 ≤ d < n`), together with the counter value that was actually used.
 *
 * Exposed so the chain builder can:
 *  - call with `counterStart = 0` for the normal path, and
 *  - call with `counterStart = previousCounter + 1` when the resulting tweak
 *    `t = TaggedHash(...)` is itself ≥ n (spec "Succession tweak": "If
 *    `t[i] ≥ n` it is invalid; re-derive `P_internal[i]`"). Re-deriving means
 *    stepping past the counter that produced the offending internal key.
 *
 * Encoding rule (matches the spec's Python reference):
 *  - counter == 0  →  info = "internal-key" || u32be(generation)
 *  - counter > 0   →  info = "internal-key" || u32be(generation) || u8(counter)
 */
internal fun deriveScalarAtCounter(rootSecret: ByteArray, generation: Int, counterStart: Int): Pair<ByteArray, Int> {
    require(rootSecret.size == 32) { "root_secret must be 32 bytes, got ${rootSecret.size}" }
    require(generation >= 0) { "generation must be >= 0, got $generation" }
    require(counterStart in 0..255) { "counterStart must be in 0..255, got $counterStart" }

    val baseInfo = "internal-key".encodeToByteArray() + u32be(generation)

    var counter = counterStart
    while (counter <= 255) {
        val info = if (counter == 0) baseInfo else baseInfo + byteArrayOf(counter.toByte())
        val d = hkdfSha256(rootSecret, NIP41_HKDF_SALT, info, 32)
        if (Secp256k1.secKeyVerify(d)) return d to counter
        counter++
    }
    error(
        "internal seckey derivation exhausted 256 counter values from $counterStart " +
            "(should be infeasible at ~256 × 2^-128 probability)",
    )
}

/** Big-endian 32-bit encoding. */
internal fun u32be(value: Int): ByteArray = byteArrayOf(
    ((value ushr 24) and 0xff).toByte(),
    ((value ushr 16) and 0xff).toByte(),
    ((value ushr 8) and 0xff).toByte(),
    (value and 0xff).toByte(),
)

File: harness/nip41/derivation.py
"""The identity chain's per-generation internal keys."""

from vendor.bip340_reference import n, G, point_mul
from nip41.hkdf import hkdf_sha256

SALT = b"nip41-key-rotation-v1"
INFO_PREFIX = b"internal-key"


def info_for(i: int, attempt: int) -> bytes:
    """HKDF `info` for generation `i`. attempt 0 = no suffix; k>=1 appends u8(k)."""
    info = INFO_PREFIX + i.to_bytes(4, "big")
    if attempt > 0:
        info += bytes([attempt])
    return info


def internal_scalar(root_secret: bytes, i: int) -> int:
    """Internal private scalar for generation `i`, reduced into 1..n-1."""
    attempt = 0
    while True:
        okm = hkdf_sha256(root_secret, SALT, info_for(i, attempt), 32)
        d = int.from_bytes(okm, "big")
        if 0 < d < n:
            return d
        attempt += 1  # reject-and-rederive (probability ~2**-128)


def internal_key(root_secret: bytes, i: int):
    """Returns (scalar, point) for generation `i`."""
    d = internal_scalar(root_secret, i)
    return d, point_mul(G, d)
